### PR TITLE
DP-75 Publish to private repo

### PR DESCRIPTION
**Pull request by simonwgill**
Draft for the moment, since I want to see how it interacts with Jenkins.

@zhelyan I've added you as a reviewer so you can see what I'm trying to achieve so far.

### CODE REVIEW

**Comment by zhelyan** [comment]
I like the concept but I think we need to wrap it up in a standalone script / task dealing with the version bumping / naming. Otherwise you'll have to duplicate the logic in every repo and rely 100% on Jenkins to do the releases.
Tools like e.g.  https://github.com/svenfuchs/gem-release can provide a starting point.

Pushing the gems also affects the release process  so it would be useful to go back to it and answer some questions first:

- what should and should not be released. Are we publishing PR versions?
- what's involved in doing a release ( version bump, commits, changelog, pushing, notifications, readme? etc)
- who's doing the releases
- naming convention
- do we need pre-releases and if yes - how long should pre-releases stick around in the gem server
- where are we publishing
- how do we do a release
...

**Comment by simonwgill** [comment]
A build on master would become a full release, and that uses the version number as it stands in the repository. If it's just that, it's pretty simple, because jenkins can call rake build, then the gem nexus call can just use pkg/*.gem

Prereleases is a weird one though - you've got to pick one of the main three components to bump when you package it, otherwise the actual release of that version gets picked up by dependendency management. That's why it's faking the release version through the rakefile. There's something of a problem with the way I've done this though - it won't push the prerelease artifact to nexus because the gemspec version and the filename version don't match.

I'd still like us to be able to use prerelease versions at least from the develop branch, but I'm still trying to figure out how to handle an incrementing build number for the prerelease identifier.

Agree on moving the logic into one place. Needs proving somewhere first though, which is why I picked this one.

I've added some pros and cons about github and nexus to https://citizensadvice.atlassian.net/wiki/spaces/OPS/pages/1671299091/Managing+releases+using+a+private+gem+server It turns out that it's easier to push and pull with nexus than github.

File: Jenkinsfile [hunk 1]
```diff
@@ -45,5 +45,19 @@ node('docker && awsaccess') {
...
+    stage('Push to repository') {
+      def secrets = [
+          [$class: 'VaultSecret', path: 'secret/devops/sonatype_nexus', secretValues: [
+              [$class: 'VaultSecretValue', envVar: 'NEXUS_USER', vaultKey: 'username'],
+              [$class: 'VaultSecretValue', envVar: 'NEXUS_PASSWORD', vaultKey: 'password']]]
+      ]
+
+      if (env.BRANCH_NAME == 'master') {
+        wrap([$class: 'VaultBuildWrapper', vaultSecrets: secrets]) {
+          sh("rake release")
```
Comment (simonwgill) [inline Jenkinsfile:52]: The nexus repository currently allows overwriting of artifacts. Not 100% sure that's the right behaviour we want right now.

**Comment by simonwgill** [comment]
@zhelyan I've stripped this back so it's just pushing artifacts when master gets merged and doing so by overriding the push subtask of the rake release task provided by bundler. Much cleaner, and doesn't have any weird version number shenaniganery in it.

File: Jenkinsfile [hunk 2]
```diff
@@ -45,5 +45,19 @@ node('docker && awsaccess') {
         reportName: 'cloudformation-ruby-dsl documentation'
       ])
     }
+
+    stage('Push to repository') {
+      def secrets = [
```
Comment (zhelyan) [inline Jenkinsfile]: just a personal preference - you can use the much more concise https://github.com/citizensadvice/jenkins-global-pipeline-lib/blob/master/vars/withVaultSecrets.txt

Comment (simonwgill) [reply Jenkinsfile]: I'd forgotten about that. Easy to fix.

Any other issues before I widen out the reviews?

Comment (zhelyan) [reply Jenkinsfile]: looks good to me. The only thing I would try is to leverage some of the internal methods here https://github.com/rubygems/bundler/blob/35be6d9a603084f719fec4f4028c18860def07f6/lib/bundler/gem_helper.rb#L98

e.g. `sh_with_input`

Comment (zhelyan) [reply Jenkinsfile]: ..which is protected. :-))

**Review by zhelyan — DISMISSED** [review]

**Review by ruthwells — APPROVED** [review]
We talked through it - it all seems reasonable, even if I couldn't reproduce it from scratch!

**Review by zhelyan — APPROVED** [review]

**Review by michelesr — APPROVED** [review]

**Pull request merged**